### PR TITLE
log4cxx: Add version 1.2.0 and replace duplicate version 0.12.0 with version 0.13.0

### PR DESCRIPTION
**Pull request by swebb2066**
log4cxx/1.0.0 is a major release that fixes a number of long-standing issues and is designed to make Log4cxx ABI stable moving forward.

For a full list of changes see the [release notes](https://logging.apache.org/log4cxx/latest_stable/changelog.html#rel_1_0_0)

log4cxx  is dependent on  PR #16142 [apr-util: conan v2 support](https://github.com/conan-io/conan-center-index/pull/16142)
---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [ ] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by prince-chrismc** [comment]
Make sure to comment in #4 for acess

**Comment by prince-chrismc** [comment]
You also need to sign the CLA https://github.com/conan-io/conan-center-index/pull/16679#issuecomment-1482355536 :)

**Comment by prince-chrismc** [comment]
Doing a quick scan this looks super good, keep it up!

**Comment by prince-chrismc** [comment]
>2 out of 3 committers have signed the CLA.
✅ stephen-webb
✅ swebb2066
❌ Stephen Webb

File: recipes/log4cxx/config.yml
Comment (SpaceIm) [inline recipes/log4cxx/config.yml:9]: keep previous ordering please

File: recipes/log4cxx/all/patches/0002-fix-expat-target.patch
Comment (SpaceIm) [inline recipes/log4cxx/all/patches/0002-fix-expat-target.patch:1]: this patch can be avoided with set_property

File: recipes/log4cxx/all/conanfile.py
Comment (SpaceIm) [inline recipes/log4cxx/all/conanfile.py:72]: version range is not allowed for the moment, except for openssl & cmake

File: recipes/log4cxx/all/conanfile.py
Comment (SpaceIm) [inline recipes/log4cxx/all/conanfile.py:12]: the convention in CCI is <RecipeName>Conan, not super important.

File: recipes/log4cxx/all/conandata.yml
Comment (SpaceIm) [inline recipes/log4cxx/all/conandata.yml:10]: Do you really need these 2 versions? Since there is 1.1.0, and there are many patches in 1.0.0 & 0.13.0, I'm not sure it's worth the effort and maintenance burden.

Comment (swebb2066) [reply recipes/log4cxx/all/conandata.yml]: Version 1.1.0 is sufficient - 0.13.0 has a different ABI and fixes some 0.12.0 mutli-threading performance issues.

File: recipes/log4cxx/all/patches/0002-fix-expat-target.patch
Comment (swebb2066) [reply recipes/log4cxx/all/patches/0002-fix-expat-target.patch:1]: Can you tell me how? I tried:
```
  tc = CMakeDeps(self)
  if Version(self.version) < "0.13.0":
      tc.set_property("expat::expat", "cmake_target_name", "EXPAT::EXPAT")
  tc.generate()
```

File: recipes/log4cxx/all/conanfile.py
Comment (swebb2066) [reply recipes/log4cxx/all/conanfile.py:72]: Why? - it seems to work

Comment (swebb2066) [reply recipes/log4cxx/all/conanfile.py:72]: I notice other libraries are now using version ranges.

**Comment by jabbas** [comment]
@swebb2066 Hey Steven, are you still working on this?

**Comment by prince-chrismc** [comment]
@jabbas Dont wait, feel free to continue this work in your own pull request :)

**Comment by swebb2066** [comment]
I have done all I need to do - I am patiently waiting for someone to answer my question or accept the PR.

**Comment by jabbas** [comment]
@SpaceIm @franramirez688 how we can make the Conan v2 pipeline to be able to build the missing apr/1.7.0?

**Comment by uilianries** [comment]
I just triggered a new CI build to provide `apr/1.7.0`, so we can try Conan 2.x again.

File: recipes/log4cxx/all/conanfile.py
Comment (uilianries) [inline recipes/log4cxx/all/conanfile.py:64]: Please, avoid using version ranges for packages that are not allowed yet: https://github.com/conan-io/conan-center-index/blob/master/docs/adding_packages/dependencies.md#version-ranges

Using version range does not guarantee compatibility between versions, it depends on each project. We are preparing a list of new packages allowed in CCI, but so far only `openssl` and `cmake` are allowed.

File: recipes/log4cxx/all/conanfile.py
Comment (prince-chrismc) [inline recipes/log4cxx/all/conanfile.py:146]: Why are these being removed?

File: recipes/log4cxx/all/conanfile.py
Comment (prince-chrismc) [inline recipes/log4cxx/all/conanfile.py:114]: Why are you removing these?

File: recipes/log4cxx/all/patches/0002-find-apr.patch
Comment (prince-chrismc) [inline recipes/log4cxx/all/patches/0002-find-apr.patch:13]: This can be fixed in the recipe with CMakeDeps and set_property. The patch is not required

File: recipes/log4cxx/all/patches/0002-fix-expat-target.patch
Comment (prince-chrismc) [reply recipes/log4cxx/all/patches/0002-fix-expat-target.patch:1]: You need to use the Conan name of the recipe or its components, not the default cmake target

File: recipes/log4cxx/all/patches/0002-find-apr.patch
Comment (swebb2066) [reply recipes/log4cxx/all/patches/0002-find-apr.patch:13]: Can you tell me how? I tried:

  tc = CMakeDeps(self)
  if Version(self.version) < "0.13.0":
      tc.set_property("expat::expat", "cmake_target_name", "EXPAT::EXPAT")
  tc.generate()

File: recipes/log4cxx/all/conandata.yml
Comment (prince-chrismc) [inline recipes/log4cxx/all/conandata.yml:35]: Version conflicts should be solved in upstream recipes, adding patches to change the behavior is a bit much.

`apr-util` should get it's own PR and bump deps are merged automatically so they are very easy

File: recipes/log4cxx/all/conanfile.py
Comment (prince-chrismc) [inline recipes/log4cxx/all/conanfile.py:88]: is there a reason to only output info? if the user incorrectly configures the graph it should raise an error

File: recipes/log4cxx/all/conanfile.py
Comment (prince-chrismc) [inline recipes/log4cxx/all/conanfile.py:52]: Why was this removed? it seems to be explicitly linked against in the build scripts you are patching

File: recipes/log4cxx/all/patches/0002-find-apr.patch
Comment (prince-chrismc) [reply recipes/log4cxx/all/patches/0002-find-apr.patch:13]: https://docs.conan.io/2/reference/tools/cmake/cmakedeps.html?highlight=set_property#overwrite-properties-from-the-consumer-side-using-cmakedeps-set-property

https://github.com/conan-io/conan/blob/048295bd4bcc6be9692dab92e232248b67f9541c/conans/test/integration/toolchains/cmake/cmakedeps/test_cmakedeps.py#L304

```py
def generate(self):
    deps = CMakeDeps(self)
    deps.set_property("expat", "cmake_target_name", "EXPAT::EXPAT")
    deps.generate()
```

File: recipes/log4cxx/all/conanfile.py
Comment (prince-chrismc) [inline recipes/log4cxx/all/conanfile.py:52]: Is there any references for this?

File: recipes/log4cxx/all/patches/0002-find-apr.patch
Comment (prince-chrismc) [reply recipes/log4cxx/all/patches/0002-find-apr.patch:13]: This can be used here for apr and apr-utils not only expat. 
This patch is not required

File: recipes/log4cxx/all/patches/0002-make-odbc-optional.patch
Comment (prince-chrismc) [inline recipes/log4cxx/all/patches/0002-make-odbc-optional.patch:11]: We should not be adding options to build scripts 🤔 this should be removed one the upstream recipe is fixed

File: recipes/log4cxx/all/patches/0004-find-apr.patch
Comment (prince-chrismc) [inline recipes/log4cxx/all/patches/0004-find-apr.patch:10]: This can be fixed in the recipe itself

**Comment by jabbas** [comment]
@swebb2066 awesome work! Much appreciated!

File: recipes/log4cxx/all/conanfile.py
Comment (uilianries) [inline recipes/log4cxx/all/conanfile.py:154]:
```suggestion
        rmdir(self, os.path.join(self.package_folder, "share"))
        rmdir(self, os.path.join(self.package_folder, "lib", "pkgconfig"))
        rmdir(self, os.path.join(self.package_folder, "lib", "cmake"))
```
Let's simplify, rmdir does not raise an error when a folder does not exist.

**Review by uilianries — COMMENTED** [review]
Thank you for your contribution, I added some comments about the PR.

File: recipes/log4cxx/all/conandata.yml
Comment (uilianries) [reply recipes/log4cxx/all/conandata.yml:16]: Up!

**Comment by valgur** [comment]
@RubenRBS You might want to review this. The build passed but `pr-merge` seems to have failed for some reason.

File: recipes/log4cxx/all/conanfile.py
Comment (valgur) [inline recipes/log4cxx/all/conanfile.py:49]: An `options_description` attribute would not hurt. E.g. it's not immediately particularly clear what the purpose of `char_type` vs `with_wchar_t` is.

File: recipes/log4cxx/all/conanfile.py
Comment (valgur) [reply recipes/log4cxx/all/conanfile.py:72]: Would be nice to have a reference for this, but it only affects outdated versions and it's a new option, so I'm fine with it as is.

File: recipes/log4cxx/all/conanfile.py
Comment (valgur) [inline recipes/log4cxx/all/conanfile.py:102]:
```suggestion
                self.output.info(f"Version {self.version} requires C++17. log4cxx version 1.x does not.")
```

File: recipes/log4cxx/all/conanfile.py
Comment (valgur) [inline recipes/log4cxx/all/conanfile.py:108]:
```suggestion
```

File: recipes/log4cxx/all/conanfile.py
Comment (valgur) [inline recipes/log4cxx/all/conanfile.py:128]:
```suggestion
        tc.variables["BUILD_TESTING"] = False
        tc.variables["LOG4CXX_INSTALL_PDB"] = False
```
log4cxx/0.12 sets CMake policy version to 3.13, so `tc.variables` can be used safely.

**Review by valgur — APPROVED** [review]
The dependencies probably need a version bump, otherwise LGTM.

**Review by valgur — APPROVED** [review]

**Review by danimtb — APPROVED** [review]

**Pull request merged**
